File: leetcode/dp/2d/0005_longest_palindromic_substring.py
# LC 5. Longest Palindromic Substring | Medium
class Solution:
    # Expand around center - O(n^2) time, O(1) space
    def longestPalindrome(self, s):
        res = ""
        for i in range(len(s)):
            for l, r in [(i, i), (i, i+1)]:
                while l >= 0 and r < len(s) and s[l] == s[r]: l -= 1; r += 1
                if r - l - 1 > len(res): res = s[l+1:r]
        return res

File: leetcode/dp/2d/0005_longest_palindromic_substring.cpp
// LC 5. Longest Palindromic Substring | Medium
#include <string>
using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        int start = 0, maxLen = 0, n = s.size();
        auto expand = [&](int l, int r) {
            while (l >= 0 && r < n && s[l] == s[r]) { l--; r++; }
            if (r-l-1 > maxLen) { start = l+1; maxLen = r-l-1; }
        };
        for (int i = 0; i < n; i++) { expand(i,i); expand(i,i+1); }
        return s.substr(start, maxLen);
    }
};
